File: Hard/1425. Constrained Subsequence Sum/sol.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

class Solution
{
public:
    int constrainedSubsetSum(vector<int> &nums, int k)
    {
        vector<int> dp(nums);
        deque<int> q;
        q.push_back(0);
        for (int i = 1; i < nums.size(); i++)
        {
            while (!q.empty() && q.back() < i - k)
            {
                q.pop_back();
            }
            dp[i] = max(dp[i], dp[q.back()] + nums[i]);
            while (!q.empty() && dp[q.front()] <= dp[i])
            {
                q.pop_front();
            }
            q.push_front(i);
        }
        return *max_element(dp.begin(), dp.end());
    }
};

File: Hard/1425. Constrained Subsequence Sum/sol.java
import java.util.Arrays;

public class sol {
    public static void main(String[] args) {
        // call the fn here
    }
    class Solution {
        public int constrainedSubsetSum(int[] nums, int k) {
            int[] dp = Arrays.copyOf(nums, nums.length);
            Deque<Integer> q = new ArrayDeque<>();
            q.offerLast(0);
            for (int i = 1; i < nums.length; i++) {
                while (!q.isEmpty() && q.peekLast() < i - k) {
                    q.pollLast();
                }
                dp[i] = Math.max(dp[i], dp[q.peekLast()] + nums[i]);
                while (!q.isEmpty() && dp[q.peekFirst()] <= dp[i]) {
                    q.pollFirst();
                }
                q.offerFirst(i);
            }
            int max = Integer.MIN_VALUE;
            for (int num : dp) {
                max = Math.max(max, num);
            }
            return max;
        }
    }
}